### PR TITLE
Flesh out more of the workflow, start executing PAPR on GitHub events

### PR DESCRIPTION
**Pull request by jlebon**
There's quite a bit of fixes and improvements in here. We *finally* actually start running PAPR jobs upon receiving GitHub events. The state of affairs is summed up in the last commit message:

---

The next step now is to start adapting the PAPR spawner as follow:

- parse YAML
- in `spawner`, ignore suites that are not 'container' envs for now
- in `testrunner`, migrate from `docker` to `oc` to create and monitor Kubernetes jobs

I'm thinking of doing these modifications in a new `openshift` branch of PAPR to avoid breakage in the current infra.

---

One interesting issue is job garbage-collection, since they're not deleted by default. I think for now I'll just change the spawner to only auto-clean jobs if they exited successfully.

Keeping logs of past runs is one thing Jenkins did nicely for us. But I don't want to stay tied to Jenkins for running jobs. I'd like eventually to define a queue using [custom resources](https://kubernetes.io/docs/concepts/api-extension/custom-resources/), much like how Prow does it. This will allow us to do prioritization and decentralization.

### CODE REVIEW

File: jenkins/jjb/papr.yaml [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+          wipe-workspace: false
+    properties:
+      - github: # required by GHPRB
+          url: 'https://github.com/{ghowner}/{ghrepo}'
+    triggers:
+      - github # branch pushes
+      - github-pull-request:
+
+          github-hooks: true
+          auth-id: '7a64217f-077c-4fde-a9a3-fd0ae37d4e0a'
```
Comment (cgwalters) [inline jenkins/jjb/papr.yaml:54]: Should this be public?

File: papr/papr-trigger.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                                default="")
+    branch_or_pull.add_argument('--pull', metavar='ID', default="",
+                                help="GitHub pull request ID to test")
+    return parser.parse_args()
+
+
+def generate_papr_job(args):
+    repo_name = args.repo[args.repo.index('/')+1:]
+    target_name = args.branch if args.branch else args.pull
+    uuid_name = uuid.uuid4().hex[:6] # XXX: actually check for collision
```
Comment (cgwalters) [inline papr/papr-trigger.py:49]: There has to be some out of the box solution for this somewhere.  Not a blocker now though.

File: papr/papr-trigger.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                            # XXX: pvc for git checkout caches
+                            "workingDir": "/tmp",
+                            "env": [
+                                {
+                                    "name": "IN_OPENSHIFT",
+                                    "value": "1"
+                                },
+                                {
+                                    "name": "github_repo",
+                                    "value": args.repo
```
Comment (cgwalters) [inline papr/papr-trigger.py:79]: Does this then clone the git repo again?

File: jenkins/jjb/papr.yaml (hunk 1 above)
Comment (jlebon) [reply jenkins/jjb/papr.yaml:54]: Yeah, it's just an object ID, not a credential in any way. Humans can customize those (like I did for `github-webhook-shared-secret`) but I just used what Jenkins auto-generated.

File: papr/papr-trigger.py (hunk 3 above)
Comment (jlebon) [reply papr/papr-trigger.py:79]: Right, I'm thinking of having a PV to hold cache to be shared by all the containers.

Comment (cgwalters) [reply papr/papr-trigger.py:79]: One thing that's really important for doing continuous delivery (CD as opposed to CI) is that one *needs* to keep around and track the source code to what is shipping in a rigorous fashion.  Among other reasons, for license compliance.  So as we start to think about rdgo+paci integration (actually heading towards pacd :smile: ?) this particular bit is important.

I had been tentatively thinking that we run a proper git mirror as a separate service, then other things pull from it.  This is the approach I took for gnome-continuous, and e.g. Fedora has a much crummier version of this in the lookaside cache of tarballs.

Anyways not a blocker...but definitely something to keep in mind.

Comment (jlebon) [reply papr/papr-trigger.py:79]: Yeah, that makes a lot of sense!

> (actually heading towards pacd :smile: ?)

But but, it's not as catchy! :) "pacid" ?
OK, enough hypothetical bikeshedding, heh.

Comment (cgwalters) [reply papr/papr-trigger.py:79]: (And I forgot to mention rpmdistro-gitoverlay is also structured this way)

**Pull request merged**